sdk/keyvault/azure-security-keyvault-shared/inc/azure/keyvault/shared/challenge_based_authentication_policy.hpp

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief Key Vault Challenge-Based Authentication Policy.
+ *
+ */
+
+#pragma once
+
+#include <azure/core/http/policies/policy.hpp>
+#include <azure/core/internal/credentials/authorization_challenge_parser.hpp>
+
+#include <stdexcept>
+
+namespace Azure { namespace Security { namespace KeyVault { namespace _internal {
```

Comment:
Is keyvault the only Azure service that implements challenge based authentication? Is it possible that in the future there may be other services which implement it?

Reply:
Storage may add one, also other services for which we don't have API yet support it (C# SDK have it for Azure Tables, Container Registry, ARM). It is possible that even more services support it than the current SDKs have it implemented for.

Reply:
If that's the case, does it make sense for the challenge based authentication policy to move to core and out of identity?

Reply:
You mean out of KeyVault? Every service looks for different challenge parameters. If you look at other SDKs, each service implements slightly different challenge-based authentication policy. The goal of Core is to provide enough flexibility in `BearerTokenAuthenticationPolicy` so that each service can customize it. And that is what is done to `BearerTokenAuthenticationPolicy` in this PR - you can see the extension points do provide the same functionality as C# SDK, for example.
But what if we find out that the same policy can be largely reused by different services? I don't think it will happen, but if it will - then we can move it to core, and the existing pre-service policies (`KeyVault::_internal::ChallengeBasedAuthPolicy` in this case) will become typedefs to the new `Core::_internal::ChallengeBasedAuthPolicy`. Either way, they both live in `_internal` namespaces, so we have more flexibility there.

Reply:
If this is truly keyvault specific, then the policy name shouldn't be `ChallengeBasedAuthPolicy`, it should be `KeyVaultChallengeBasedAuthPolicy` to express the fact that this is a KeyVault specific policy (the namespace isn't sufficient IMHO - it's often removed by `using` directives, and by putting the service into the name, you make it abundantly clear that this is a policy for the keyvault flavor of challenge based authentication. 

There's an existing pattern that matches this with the Storage `SAS` types - there's a `BlobSasBuilder`, a `AccountSasBuilder`, etc - the service name is embedded in the type name, not just the namespace.

Reply:
I agree, good suggestion, I renamed it to `KeyVault::_internalKeyVaultChallengeBasedAuthenticationPolicy`, as well as the header file to `keyvault/shared/keyvault_challenge_based_authentication_policy.hpp`